stackit/internal/services/ske/cluster/resource.go

```diff
@@ -848,7 +919,139 @@ func toNodepoolsPayload(ctx context.Context, m *Model) ([]ske.Nodepool, error) {
 		}
 		cnps = append(cnps, cnp)
 	}
-	return cnps, nil
+	return cnps, deprecatedVersionsUsed, nil
+}
+
+// latestMatchingMachineVersion gets the latest machine image version to be used in the create/update payload
+// it's determined based on the available versions for the input OS (machineName), the minimum version configured by the user and the current version in the cluster
+// if the minimum version is not set, get the current one (if exists) or the latest for the input OS
+// if the minimum version is set but it is a downgrade, use the current version
+// for the minimum version selected, if a patch is not specified, get the latest patch for that minor version
+// for the version returned, check the state and return if it is deprecated or not
+func latestMatchingMachineVersion(availableMachineImages []ske.MachineImage, machineVersionMin *string, machineName string, currentMachineImage *ske.Image) (version *string, deprecated bool, err error) {
+	deprecated = false
+
+	if availableMachineImages == nil {
+		return nil, false, fmt.Errorf("nil available machine versions")
+	}
+
+	var availableMachineVersions []ske.MachineImageVersion
+	for _, machine := range availableMachineImages {
+		if machine.Name != nil && *machine.Name == machineName && machine.Versions != nil {
+			availableMachineVersions = *machine.Versions
+		}
+	}
+
+	if len(availableMachineImages) == 0 {
+		return nil, false, fmt.Errorf("there are no available machine versions for the provided machine image name %s", machineName)
+	}
+
+	if machineVersionMin == nil {
+		// different machine OS have different versions
+		// so if the current machine is nil or the machina name was updated,  get the latest one
+		// otherwise, the current machine version should be used
+		if currentMachineImage == nil || currentMachineImage.Name == nil || *currentMachineImage.Name != machineName {
+			latestVersion, err := getLatestSupportedMachineVersion(availableMachineVersions)
+			if err != nil {
+				return nil, false, fmt.Errorf("get latest supported machine image version: %w", err)
+			}
+			return latestVersion, false, nil
+		}
+		machineVersionMin = currentMachineImage.Version
+	} else if currentMachineImage != nil && currentMachineImage.Name != nil && *currentMachineImage.Name == machineName {
+		// For an already existing cluster, if os_version_min is set to a lower version than what is being used in the cluster
+		// return the currently used version
+		machineVersionUsed := *currentMachineImage.Version
+		machineVersionMinString := *machineVersionMin
+
+		minVersionPrefixed := "v" + machineVersionMinString
+		usedVersionPrefixed := "v" + machineVersionUsed
+
+		if semver.Compare(minVersionPrefixed, usedVersionPrefixed) == -1 {
+			machineVersionMin = currentMachineImage.Version
+		}
```

Comment:
```suggestion
        // If the os_version_min is set but is lower than the current version used in the cluster,
        // retain the current version to avoid downgrading.
		minimumVersion := "v" + *machineVersionMin
		currentVersion := "v" + *currentMachineImage.Version

		if semver.Compare(minimumVersion, currentVersion) == -1 {
			machineVersionMin = currentMachineImage.Version
		}
```